### ISSUE TITLE
Enhance Multivariate MODE to generate object statistics for each input field requested by the user.

### ISSUE THREAD

**Issue by dwfncar** [dtcenter/MET#1283]
## Describe the Enhancement ##
I am repurposing this existing issue to better define the multi-variate MODE enhancements that are needed.

1. While the existing `MODEMultivarConfig_default` file exists in the repository, it is NOT actually read by Multi-Variate MODE! In all cases, MODEConfig_default is read instead. Need to understand why, and either update MODE to read it in the right circumstances or just remove it from the repo.

2. If `MODEMultivarConfig_default` is retained, modify the `nc_pairs_flag` setting in it to specify it as a dictionary, just like it is specified in MODEConfig_default. This is needed to make the logic of the MODE wrapper clear and consistent.

3. Make large updates to the multi-variate logic, as described below.
- As of MET version 10.1.0, multi-variate MODE generates output NetCDF files named `f_super.nc` and `o_super.nc`. However, it does not actually write out any object statistics.
- Recommend that it continue to write out the super object information, but write out both fields to the standard MODE NetCDF output file instead (i.e. *_obj.nc).
- Add a new config option that is an array of booleans with the same length as the number of forecast and observation fields: `multivar_intensity_flag = [ TRUE, TRUE, FALSE ];` (default will be FALSE for all)
- After generating the super objects, loop through the `multivar_intensity_flag` booleans:
   - If all `multivar_intensity_flag` booleans are false, run MODE once on the binary 0/1 super objects to generate geometric object statistics and write them to the output "*_obj.txt" file.
   - Otherwise, for each `multivar_intensity_flag` boolean flag that is true, run MODE once substituting in the raw intensity values for the corresponding fcst/obs fields. The object definition has already been performed, but we're just regenerating the statistics using the intensities for those fields. Write the output to the "*_obj.txt" output file. For any FALSE booleans, the intensity statistics will contain missing values.
   - So the "*_obj.txt" may now contain output for multiple fields with different intensity values and potentially total interest values. Need to differentiate between them in the output in the FCST_VAR and OBS_VAR columns. Recommend including the forecast and observation field names in those output columns.
   - Be sure to check the `nc_pairs_flag` option and write intensity values to the NetCDF file, as requested by the user.
   - Add a new config option that allows the user to change the forecast and observation field names. Currently the super object output name is ‘super’, but users may be running multivariate MODE on multiple complex events (dryline/blizzard), and it will be beneficial to have a way to differentiate between the output
      - Add `multivar_name="";` to change the name of the multivariate mode output as listed in the ascii file output, for example. (default will be "super")
      - For example, if we define `multivar_name = "blizzard";` and we request intensity output for visibility, our ascii file(s) should append this to the FCST_VAR and OBS_VAR names e.g. blizzard_vis/vis_blizzard (preferred order?)
         - If intensity isn’t requested for any field, the name would be blizzard for the geometric object statistics?
- Other considerations
   - Determine whether there should be separate MODE output files for each field where intensity statistics are requested, or all output contained in a single file
      - Do what makes sense, when thinking about parallelization
   - No need to keep mode runs from the individual fields - cuts down on processing time
   - Keep this info in memory for continuing the mvmode process
 
### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2700042 WPC

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
Related to MET Issues:
- #1184 
- #1282 
- #1284 

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

**Comment by lindsayrblank** [comment]
Replace system calls with function calls.

**Comment by JohnHalleyGotway** [comment]
Please note that any changes to how Multi-Variate MODE should be run based on this issue should be noted on the METplus issue for the Multi-Variate MODE wrapper (dtcenter/METplus#1585).

**Comment by davidalbo** [comment]
Meeting notes, for reference:    https://docs.google.com/document/d/15PUUn8VoI8-3tzlr56NiFIxxF7rSykO8_aEJalqa8jg/edit

**Comment by davidalbo** [comment]
@JohnHalleyGotway an engineering question:
It looks like these get compiled but never used:
```
mode_multivar.cc
mode_singlevar.cc
```
Both of them look a lot like they are formerly or future main() candidates.

Instead we have these, both called from mode.cc: main().
```
mode_frontend.cc
multivar_frontend.cc
```

Correct? 
Are the former two still in the source directory as some kind of work in progress, or a partial attempt to turn mode into two programs single and multivar?

**Comment by davidalbo** [comment]
@hertneky do you think anyone will want support for **multiple** convolution radii and/or thresholds for the multivariate use that I'm working on? The single mode design supports this so it could be added.   I do anticipate some tricky complexity, however.  Example: If field 1 has 2 thresholds, field 2 has 3 thresholds, and field 3 has 1, what would the output be?  You could generate 6 sets of super objects for all the different threshold combinations, and... then what?  

Simpler at least to start would be to disallow multiple radii/thresholds for multivariate situation.
As a scientist, what are your thoughts on this?  As a multiple choice:
Nice but can live without it
Needed for sure
Not needed for sure

Thanks
Dave

**Comment by hertneky** [comment]
@davidalbo As you mentioned this is already a capability of MODE and is controlled by the "quilt" option. I think it could be useful to allow permutations. These permutations would only apply wrt identifying the super object (not the individual mode runs).
When I ran multivariate MODE for identifying blizzard objects, I ran it multiple times with various convolution thresholds for a couple of the fields to figure out what thresholds were needed to identify the object of interest. It could have been quicker if permutations were allowed, so I can see it being useful.
In the output, obj.txt for example, there could be rows pertaining to the first permutation, then the 2nd, then the 3rd, etc. Not currently sure how that would look for the netcdf output. I have never actually run MODE with quilting set as True, so I am not sure what the output looks like for regular MODE. @JohnHalleyGotway likely knows or I can do a quick run.
If you want to get the initial mvmode capability working without allowing quilting first, and then adding the capability, I think that would be fine if it makes sense in an engineering perspective.

**Comment by davidalbo** [comment]
@hertneky Noted.  Given how complex this is without the 'quilting, I'll implement a version that error exits if it is multivariate situation with quilting configured first,  so I can learn more about output details without that, then add it in later.

**Comment by davidalbo** [comment]
I'm making progress.  I have a crude version implemented that is creating the super objects, and then masking the original inputs using those objects, and finally running mode on each of the masked inputs to get statistics.  I have not yet implemented the configuration where you can select a subset of the inputs to do statistics on.  This is really rough at this point.

**Comment by davidalbo** [comment]
Hi @hertneky.   I'm going deep into the config parsing, and can see that a line like this in your setup:
 ` MODE_MULTIVAR_LOGIC = #1 && #2 && #3 `  
causes obs and fcst to use the same logic.
In the source code the same kind of line can also can be put inside of the fcst and obs config areas independently, in which case fcst and obs each could have it's own logic.  It looks like metplus does not support this, i.e. there is no way to set it up from metplus. Do you know?  If not maybe it should.  Something like:

```
MODE_FCST_MULTIVAR_LOGIC
MODE_OBS_MULTIVAR_LOGIC
```

to override

`MODE_MULTIVAR_LOGIC`

This does seem to open it up to where obs and fcst could have different number of inputs and different logic, which is more general.

**Comment by hertneky** [comment]
@davidalbo While this is an interesting idea and it does open up flexibility in defining a super object, it makes it very complex with respect to intensities and perhaps not possible to output intensities depending on inputs. As in the issue, we discussed having a check to ensure the number of booleans in the multivar_intensity_flag matched the number of fcst/obs inputs.

For example, if FCST `multivar_logic = "#1 && #2 && #3";` and OBS `multivar_logic = "#1 && #2";` then how does the multivar_intensity_flag work with 3 variables in the fcst and 2 in the obs? I think `multivar_intensity_flag = "true, true, false"` would error, or maybe it would ignore the 3rd flag. 

If you want to talk about this more, then we can meet with the larger group to discuss.

**Comment by davidalbo** [comment]
@JohnHalleyGotway @hertneky please provide some feedback on the algorithm as it now stands (below (I'm leaving out some smaller details)).  I think I can take some of this out, i.e. don't need to do all these steps.  Basically I'm running the **complete** mode algorithm for both the first (full data) pass and the second (masked data) pass.


```

for each input
     read in the obs and fcst data and do things to it if configured (masking, percentile thresholds, ... )
     do the convolution step
     compute contingency table stats (if configured)
     do merging
     do matching
     produce all the outputs into the numbered subdirectory (01, 02, ..):
              *_cts.txt, *_obj.txt, *.ps, *.nc
     keep what is needed in memory (fcst and obs raw data, object data, ...) for use below:

create super objects from the fcst/obs object data using the configured rules

for each input
    if multivar_intensity_flag is true
          mask the raw data to be only within the super objects
          use this masked raw data for these steps:
              do the convolution step
              compute contingency table stats (if configured)
              do merging
              do matching
              produce all the outputs into the numbered subdirectory (01, 02, ..): 
                 (note: overwrites what was created before)
                 *_cts.txt, *_obj.txt, *.ps, *.nc

Write superobject netcdf output files.

```

**Comment by hertneky** [comment]
In first for-loop
>             produce all the outputs into the numbered subdirectory (01, 02, ..):
>               *_cts.txt, *_obj.txt, *.ps, *.nc

If I understand this correctly, you are saving off the individual MODE runs here. I recall thinking these individual MODE runs weren't necessary to output and that the objective is only the output for the super-object. Writing the individual MODE runs would also take longer.

In 2nd for-loop
>              produce all the outputs into the numbered subdirectory (01, 02, ..): 
>                (note: overwrites what was created before)
>                *_cts.txt, *_obj.txt, *.ps, *.nc
Does there need to be a numbered subdirectory since the mode output has the lead/date/valid time appended? Hmm, maybe since I am not sure yet how the f_super.nc/o_super.nc will be saved - assuming in the *.nc though.

**Comment by davidalbo** [comment]
For the first loop, I have now tried doing everything except creating outputs, and that seems to work. What is needed for part 2  is the objects (from part 1) which are used to create super objects, and the raw data intensities for the input fields.  I'm keeping all that in memory.

For the 2nd loop, the question in my mind is what is the output that is wanted, and how do I implement that, and how do we represent that output. It's described as 'statistics', I need to expand on what that means. We can get into this at tomorrows tag up.   Because each field for which we compute these statistics shares the same super objects, maybe you are right, no numbered subdirs are needed.  It's here where I need some guidance.

**Comment by davidalbo** [comment]
Things not yet explored or tested as regards the newer design:

- shift right
- masking in the config
- contingency table statistics
- clusters
- polyline netCDF output
- quilting (multiple radii/thresholds)

**Comment by davidalbo** [comment]
Notes from the tag up while fresh:

- [x] first pass, no merging or matching, only simple objects needed
- [x] second pass, do the full mode run on masked inputs, rename output files to include the magic string (var_info class)
- [x] if the merge_flag=THRESH and the logic has a union (or), then print a warning.
- [x] Hold off on quilting
- [x] Move the multivar intensity lookup_bool_array method to be with other lookup stuff
- [ ]

**Comment by davidalbo** [comment]
I've now got output files with longer names including 'magic string' values.  Because obs and forecast might be different I put in both, so the names are long, maybe too long. Examples below.  

@hertneky what do you think looks best and works for you?   Or suggest something else, I'm open to anything.

New long names:

```
mode_Fcst_VIS_L0_Obs_VIS_L0_210000L_20210201_210000V_000000A_cts.txt
mode_Fcst_WIND_Z10_Obs_WIND_Z10_210000L_20210201_210000V_000000A_cts.txt
```

I could shorten to just have the forecast magic string (without the "Fcst):

```
mode_VIS_L0_210000L_20210201_210000V_000000A_cts.txt
mode_WIND_Z10_210000L_20210201_210000V_000000A_cts.txt
```

and could even leave off the L0/Z10 information:

```
mode_VIS_210000L_20210201_210000V_000000A_cts.txt
mode_WIND_210000L_20210201_210000V_000000A_cts.txt
```

**Comment by JohnHalleyGotway** [comment]
@davidalbo and @hertneky another option would be comparing the magic strings for the fcst and obs. If they're the same, only include it once. If they differ, include both.

That'd make *parsing* the file name more difficult because the number of *elements* could differ based on the configuration. But people shouldn't be parsing metadata out filenames anyway. So perhaps that's not a concern?

**Comment by davidalbo** [comment]
For the case where multivar_intensity = [FALSE, FALSE, FALSE], what is wanted?  As described in the initial issue or has something evolved?  From above:

- As of MET version 10.1.0, multi-variate MODE generates output NetCDF files named f_super.nc and o_super.nc. However, it does not actually write out any object statistics.
- Recommend that it continue to write out the super object information, but write out both fields to the standard MODE NetCDF output file instead (i.e. *_obj.nc).
- If all multivar_intensity_flag booleans are false, run MODE once on the binary 0/1 super objects to generate geometric object statistics and write them to the output "*_obj.txt" file.

**Comment by hertneky** [comment]
@davidalbo If all are false, then the super objects are compared directly (no need to bring in raw intensity fields). Then the intensity columns in the output would be set to missing values. All MODE output would still be created. The MODE file names might use a default "super" name. e.g. mode_SUPER_210000L_20210201_210000V_000000A_cts.txt

For other file names, I like John's suggestion. "If they're the same, only include it once. If they differ, include both." I am guessing that there are no METplus components that parse the file names? (e.g.METplotpy/METcalcpy/METviewer)

**Comment by davidalbo** [comment]
@hertneky For the 'all are false' case, should I assume we'd be wanting different mode algorithm settings specific to comparing super objects, which would be something like:

  conv_radius=[0];                                                                                             |                                                                              |
  conv_thresh=[ne-9999];

If so mv mode would need extra config settings.  It would use the current mvmode settings to create objects and superobjects, but then additional configuration parameters as above to compare super objects.   I think this is outside the current MODE config, so I'd need to add more config options to handle that.

**Comment by hertneky** [comment]
@davidalbo In looking at the MODEconfig, I am reminded of something that could potentially be problematic when running mvmode, the regridding within the MODE run.

My case was rather simple:
FLD1: Categorical Snow: regrid to FCST using nearest method
FLD2: Visibility and FLD3: 10m Winds both fcst/obs on same grid (no regridding required)
So defining regrid = FCST;method=nearest worked to regrid the 1 field (should have ignored the other fields since the obs are already on the same grid as fcst)

But if multiple fields need regridding, using different methods (i.e. if wind also needed regridding, one would want to use "bilinear method"). Then you could not use the regridding within the MODE call. I don't think you can currently have different regrid calls for different variables with the MODE tool, can you?

Maybe it does not matter and one should just use regrid_data_plane stand-alone for more complex regridding cases?

**Comment by davidalbo** [comment]
@hertneky I think you're right!  Looking at the code, each pair of fields gets regridded independently using the same shared regrid settings, including regrid method.  Also, what if regrid=FCST, but the forecast data for field 1 and field 2 are different grids?  Seems like that would break the superobject creation. Could that happen? It could be checked for with an error exit.  Or..  maybe we want both the ability to use different regrid methods for each field, and a single verification grid specification, maybe something like "verify to forecast grid 1" or something like that.

**Comment by hertneky** [comment]
@davidalbo Yes another good example! That would break things. So currently, you can say regrid this all to FCST FLD1, just by passing regrid = path/to/specific_gridded_fcst_file. However, you cannot specify diff methods. But, is it hard allow for different regrid methods for each field? I don't know. Maybe @JohnHalleyGotway can weigh in his expertise...

**Comment by davidalbo** [comment]
@hertneky there is always a way.  It might be a major-ish change to the configuration options.  We should tag up with @JohnHalleyGotway and talk it over.

**Comment by davidalbo** [comment]
@hertneky @JohnHalleyGotway   Another design limitation.  Current design allows exactly one file format for all forecasts and one file format for all obs. Obviously limits what you can do.

**Comment by hertneky** [comment]
@davidalbo This is not something I realized. To be sure I understand, you are talking about the forecast and obs input files for mvmode? I suppose all mine were Grib2, so I did not encounter that issue. But I could definitely see a mix of grib/netcdf.

**Comment by davidalbo** [comment]
@hertneky that's correct . 
Playing around I can add something like `file_type=GRIB2;`  to the fcst or obs, to where all the inputs share the same file_type. That comes from environment variables i.e.:  `${METPLUS_FCST_FILE_TYPE}`

 I can also add` file_type=GRIB2; `to an individual field, and it shows up in the dictionary, but in the code it is never looked for, so there is no way to set individual file_type values for each input.
I do see that some of the settings can be set both at the main fcst or obs level, but also overwritten for individual fields, for example  `merge_thresh.`

So with some work I can probably figure out a way to support setting the regridding config settings and the file_type and whatever else is needed independently for each input, along with adding something to indicate which input should be used as the template for the verification grid.

**Comment by davidalbo** [comment]
@hertneky I'm right in the code now where I'm doing this:

 for 'pass2', which is processing a field that's been masked to superobjects.

hardwire conv_thresh to ne-9999
hardwire conv_radius to 0

what about pass2 merging?  merge_flag, merge_thresh?  as it stands now it will use the values from 'pass1' for that field, unless I do something.

For our example:

 merge_thresh = <=1207.008; merge_flag = THRESH     (field2:VIS)
 merge_thresh = >=6.7056; merge_flag = THRESH         (field3:WIND)

**Comment by hertneky** [comment]
@davidalbo On pass1, we do the usual merging on the individual mode runs. Were the super objects defined from the simple objects conv_thresh or from the merge_thresh somehow (I can't recall)?

Now the question is how to approach merging on pass2 with the super objects. The same merging configs make the most sense since that is how the super objects are being defined. I am thinking whether a user might want merge thresholding on pass1, but no merging on pass2 (or vice versa). Maybe that's something we just don't allow for now. 

Edit: Also need to consider how merging is handled on the super objects if not requesting intensity (all intensity_flags set to false). I know we aren't supporting this in the first version, but needs to be considered for planning.

**Comment by davidalbo** [comment]
@hertneky In the current prototype, no merging is done on the individual mode runs, so  super objects come from the simple objects.    Merging is done in pass2 using the configured settings (for the individual inputs).

Keeping with the idea of getting something done, I'd vote for keeping it that way.....no merging in pass one, and merging in pass 2 using the configured merge settings.

I think once I figure out some way to handle the all intensity flags=false situation the merging for that case will fall out easily, because the other configurable parameters (conv radius/threshold) will need to be dealt with somehow or another anyway.

It's starting to be a lot of balls in the air, which will go into that new issue we talked about.,... a list of functionality that needs adding.

**Comment by hertneky** [comment]
@davidalbo OK, so no merging is done on the individual MODE runs so the merge_flag and merge_thresh is getting ignored.

So here is the issue I see, given that information. If WIND objects are identified using a conv_thresh>=20 m/s. then objects and super objects are defined based on the 20 m/s value. If then you set a merge_thresh>=15 m/s on the super objects, that won't do much, since you've masked the super object with the 20 m/s thresh, and basically thrown out most grid points with values below that. So I don't see how a merge_thresh would be useful.

**Comment by davidalbo** [comment]
@hertneky I see your point.  If merging were done in step one, what would that do to the super objects?  I think it means one merged object might have >1 contiguous clumps of gridpoints. Two simple objects might merge into one merged object. Am I getting it correct so far?  

But when you then create superobject and mask using the super objects for pass2, the information that they have been merged is lost, so it didn't gain you anything.   And as you say, if the merge_thresh is smaller than the conv_thresh, which I guess it really should be,  it won't do anything.  Does this mean merging will do nothing?

I suppose I could merge in the first pass, and keep track of merged objects from each input, but then when I get to the logic 1 && 2 && 3 the objects will shrink, and some remaining gridpoints might have been merged in 1 but not 2 or 3, or any combination, and maybe merged differently for each of the 3 first passes.  What would that mean?  Merged or not?   Brain overload!

**Comment by hertneky** [comment]
@davidalbo For single MODE, does the double thresholding (conv_thresh + merge_thresh) change the size of the objects identified, compared to not using merge_thresh? From what I understand of merge thresh, it defines larger objects simply for the purpose of merging, but doesn't increase the size of the resultant objects in the end. Please let me know if my understanding is wrong. If that is the case, then I think merge_thresh will do nothing, whether applied to pass1 or pass2.
Options: 
1. There is the other merge_flag=engine that can be used for merging, which I have admittedly never used.
2. merge_thresh=conv_thresh 
3. some other way....what a brain twister

**Comment by davidalbo** [comment]
@hertneky I believe you are correct, double thresholding does NOT change the size of objects, just decides which objects go together.  You might get a second opinion on that.  So far my brain overload remains where it was as regards merging.  For now we can add it to the list of things to figure out, I guess.

**Comment by hertneky** [comment]
@davidalbo I talked with John about the merge_thresh yesterday and I now see how it would be applied. When running MODE on the individual fields the merge_thresh would need to be applied and the intersecting area (if doing object intersection) of the merge_thresh would need to be saved in memory for use later for the super objects. (I can illustrate this for you if it is easier to see visually). If doing that is difficult in the code, then this option can be saved for a later enhancement, and you can have MvMODE error out if merge_thresh is set.

**Comment by davidalbo** [comment]
A visual would be good @hertneky.

**Comment by davidalbo** [comment]
@hertneky @JohnHalleyGotway .. reality check.  Is this correct?

For an individual input:
Run the mode convolution algorithm at the convolution threshold
Run mode convolution again at the merge threshold (when merge flag = BOTH or THRESH), independent of the first run.

So you could for example end up with 2 objects from the first run, but 1 from the second run (that contains both of the first run objects within it.

**Comment by hertneky** [comment]
The first part sounds good to me.

I am having trouble with the last sentence. If 2 super objects are identified in the 1st run, then you would still have 2 super objects in the 2nd run, but they may or may not be merged (identified as a cluster). Am I understanding that correctly?

**Comment by davidalbo** [comment]
![Untitled presentation](https://user-images.githubusercontent.com/16022171/236572723-303cc781-851a-4827-9910-8e1e1071e26c.png)

Hi @hertneky, This is before the superobject step, it's the individual input data..  The inner ones are the 'simple' objects (is that the right terminology?) that come from using the conv_thresh, the outer one is the simple object that comes from using a more generous threshold, the merge_thresh.  For example:

  { name="WIND"; level="Z10"; conv_radius = 5; conv_thresh = >=8.9408; merge_thresh = >=6.7056; merge_flag = THRESH; },

Two objects when >=8.9408, one object when >= 6.7056.

**Comment by davidalbo** [comment]
![Untitled presentation (1)](https://user-images.githubusercontent.com/16022171/236574676-6c4171cf-1bc8-4381-a6a9-902d2e18099c.png)

As opposed to this where the objects get bigger but do not actually mege.

**Comment by hertneky** [comment]
@davidalbo Yes these images are correct for merging. I just want to reiterate that the merge_thresh will only be used for merging. They don't make the super objects bigger in the end. The super objects are the same after the 2nd pass, but may or may not be considered part of a cluster. After the determining the super object clusters using the super_merge_thresh_object, that larger super_merge_thresh_object is not saved in any output.

**Comment by davidalbo** [comment]
@hertneky understood.  I've finally gotten the 'super_merge_thresh_object's to get created and I put in some code to output them as netCDF and look at them, and they seem reasonable.  I did run into a problem with the first input in your original example.   For the forecast the field is PrecipFlag, and you did not specify a merge_thresh or merge_flag.  The code uses a default in that case, which is merge_flag=THRESH and merge_thresh = ">=1.25".    This was causing problems because then my code tried to make bigger objects using these settings, which makes no sense. I added merge_flag=NONE explicitly for those fields, and that fixed it.  I'd have thought the default would be merge_flag=NONE, but... does not seem to be the case.

**Comment by hertneky** [comment]
@davidalbo So when merge_flag is not set, it defaults to THRESH with a value of >=1.25. Is that just for the multivariate case or does that work for single MODE as well? I would have thought the default would be NONE if not specified as well. It doesn't make sense to me to have it default to THRESH with some arbitrary value, but perhaps there was some reason behind that. I don't know.
CSNOW is binary 0/1 and PrecipFlag is categorical. So a merge thresh >=1.25 would seem problematic. We did discuss that if no merge thresh were set for a variable, then the conv_thresh would be used to create the super merge object. I see various scenarios for merge_thresh:
1. user sets merge_flag to NONE for all variables - no merging of the super objects is performed
2. user sets merge_flag to THRESH for all variables - merging of super objects is performed using merge_thresh of each field (assumes user has selected a threshold for merge_thresh for each field)
3. user sets merge_flag for some of the variables, but not all - for variables where merge_flag was not set, set it to the same as the other variables? If using THRESH, set the merge_thresh = conv_thresh for that variable

**Comment by davidalbo** [comment]
@hertneky What you say for at least 2 and 3 is what I've been implementing. Right now I'm at the exact place where I'm trying to do the merging for option 2 or 3.  I think I need to be reminded exactly what the merge algorithm should do, and how that compares to what happens for merging in the non-multivar mode situation.  This is  so I can use existing code as much as possible.  I might need a tag up, but you can try a comment if you want.

**Comment by hertneky** [comment]
@davidalbo A tag-up may be needed. So the algorithm..
1. Uses the same multivar_logic as creating the super-object(s) to create super-merge-object(s)
2. Super-objects within super-merge-objects are merged (called cluster objects)
    - For singular MODE, objects within merge-objects are merged (called cluster objects)
3. Objects not merged are simple objects
4. The remaining logic is the same for matching simple/cluster objects (if matching is requested)
The big difference is that with mvMODE, the merge_thresh of multiple variables is used to create the super-merge-object. If this doesn't answer your question and you need something more specific, we can tag-up. 
I am also on the hook for creating some diagrams for documentation purposes. I will do that next.

**Comment by davidalbo** [comment]
@hertneky is it true that simple objects produced by mode have unique integer numbers associated with them? It must be true based on the output that gets produced. I need some way to tell if a super object is within a particular super merge object. The code as is right now represents super objects as binary yes/no on a grid, so there is no information about individual objects.   If you agree that unique integers must be in there somewhere, I'll start backing up in the design and looking for those integers.

**Comment by hertneky** [comment]
Yes, to visualize this for example, one of the outputs in the nc file is fcst/obs_obj_id. These are IDs assigned to each simple object. If there are 4 objects, then the unique IDs are 1-4 (left image).
For merged objects, the output is called fcst/obs_cluster_id. Objects in a cluster due to merging will share the same ID, single objects not merged have a value of -1 (right image).
Colorbars at bottom show the values. In the right image, the large object was merged with the tiny object next to it an both objects share the ID of 1 and the bottom 2 objects were not merged and have a value of -1.

This output is from the individual MODE run for visibility. This should also be applied to the super objects.

In the object ASCII output, the IDS are also listed (to the right of the images). The 1st 4 lines correspond to the images of the FCST objects. The 1st column is the simple object IDs and the 2nd column is the clusters. Objects 1/2 were clustered (merged), objects 3/4 were not part of a cluster (denoted by CF000).

![Screenshot 2023-05-11 at 3 10 37 PM](https://github.com/dtcenter/MET/assets/39317287/cb56f751-fde5-4fce-9b5c-3940be3ba4f8)

**Comment by davidalbo** [comment]
do_fcst_convolution();
   do_fcst_thresholding();
   do_fcst_filtering();
   do_fcst_splitting();

**Comment by davidalbo** [comment]
Notes for discussion today may 30:

when merge_flag is not set, it defaults to THRESH with a value of >=1.25.  This seems wrong and should be change for both multivar mode and single var mode to... 'no merging'.   Correct?
**Answer:  Keep default as is. When it is multivariate mode, check that eaCh field has merge_flag explicitly set and error exit if not.**

For the '2nd pass' (individual inputs masked to be inside super objects), which of these steps can be skipped?
do_fcst_convolution();
do_fcst_thresholding();
do_fcst_filtering();
do_fcst_splitting();
**Solution:  Only do the splitting.  Needed to copy some data into place as is (within the engine code) to get it to work**

Did I implement this, which is also part of the 2nd pass, and is  it correct?:
Various scenarios for merge_thresh in multivar mode:
1. user sets merge_flag to NONE for all variables - no merging of the super objects is performed
2. user sets merge_flag to THRESH for all variables - merging of super objects is performed using merge_thresh of each field (assumes user has selected a threshold for merge_thresh for each field)
3. user sets merge_flag for some of the variables, but not all - for variables where merge_flag was not set, set it to the same as the other variables? If using THRESH, set the merge_thresh = conv_thresh for that variable

I'd like to talk through the expected behavior for the case where all the intensity flags are set FALSE.  It would be a mode-like run for simple objects that have no intensities, they are just binary yes/no.

**Comment by davidalbo** [comment]
@hertneky what about merging when settings are all FALSE for intensities?  It can be done and should give the same results as the 'not all FALSE' option.  Does it make sense scientifically to merge?

**Comment by hertneky** [comment]
@davidalbo It still makes sense to merge objects, even if no object intensity info is requested. The merging should be the same as if the user requested intensities.

**Comment by davidalbo** [comment]
@JohnHalleyGotway @hertneky I'm modifying the unit tests.  With intensity flags set TRUE, I'm getting output file names like the following, all in one output directory:  

`mode_Fcst_ALPHA(*,*)_Obs_ALPHA(*,*)_300000L_20120410_180000V_060000A_obj.txt
`

The unit test mode config file has:

```
fcst = {

   field = [

      {
         name  = "ALPHA";
         level = "(*,*)";
         merge_flag = NONE;
      },

```
Previously (the way the unit test is set up now)  the output files went to numbered subdirectories with file names, for example, this is the same file with the old name:

`00/mode_300000L_20120410_180000V_060000A_obj.txt `

I'm wondering about `(*,*)` in file names.

**Comment by davidalbo** [comment]
@JohnHalleyGotway eventually I'm going to want two unit tests, one with [TRUE, TRUE, TRUE] one with [FALSE, FALSE, FALSE].  The expected output files need to be put somewhere for the unit test to be able to compare them to the test output. The output file names and organization has changed, so all of that needs updating.  THis is the one piece that's not in the MET repo, is that correct?

**Comment by davidalbo** [comment]
@hertneky Could you have different multivar_logic for forecasts and for obs?  Maybe a different number of input fields for obs than forecasts?   What would that do to the intensity flags?   If intensity flag 2 and 3 are TRUE, right now it is fcst field 2 is compared to obs field 2, fcst field 3 compared to obs field 3.   Is it good enough for now to **require** equal number of fcst and obs inputs, and the same multivar_logic for fcst and obs?   And the comparisons as indicated?

**Comment by hertneky** [comment]
<h3>Release Acceptance Testing Summary</h3>

Version: METplus-5.1.0-rc1
Date: July 5, 2023
Location: cheyenne
Status (PASS/FAIL): FAIL
Description: Ran blizzard test case. Two issues: 1) When setting `MODE_MULTIVAR_INTENSITY = FALSE,TRUE,TRUE`  in the conf, the result is 'WARNING: multivar_frontend() -> empty multivar intensity array, setting to all FALSE' (in the log, I see... export METPLUS_MULTIVAR_INTENSITY="multivar_intensity = [FALSE, TRUE, TRUE];") and 2) Setting `MODE_FCST/OBS_MULTIVAR_LEVEL = L0` does not seem to replace the NA in the filename or within the FCST_LEV column in the output (e.g. mode_Fcst_bliz_NA_Obs_bliz_NA_HRRR_vs_ANALYSIS_PrecipFlag_L0_210000L_20210201_210000V_000000A_obj.txt; I do not see this being exported in the log)
Test conducted in: /glade/work/hertneky/tmp/met_rc_test

**Comment by georgemccabe** [comment]
@hertneky, we discussed this on Slack but for traceability on this thread here is what we discovered:
1) The MET variable name is actually `multivar_intensity_flag`, not `multivar_intensity` as described in the MET documentation. PR dtcenter/METplus#2236 will fix this bug.
2) It appears the rename of multivar_units to multivar_level was made after METplus 5.1.0-rc1 was created, so it did not properly set the variable. This can also be tested using the branch from the above PR.
